### PR TITLE
redis_exporter v1.0.0

### LINKED ISSUE #126: Parallelize requests
It looks like the exporter queries each redis node sequentially, would it be possible to parallelize these requests?

### LINKED ISSUE #178: Export counters as counters
I work on the [Stackdriver integration with Prometheus](https://github.com/Stackdriver/stackdriver-prometheus) and a user filed [an issue](https://github.com/Stackdriver/stackdriver-prometheus/issues/14#issue-343468080) because this redis exporter exports counters as gauges, and Stackdriver doesn't allow rate computations over gauge. I know this isn't a major problem for Prometheus (it can mostly cause glitches on process restart), but would you consider fixing the [setMetrics method](https://github.com/oliver006/redis_exporter/blob/master/exporter/redis.go#L782) to set the right metric type? Any workaround on our side will be hard to maintain. Thanks!

### LINKED ISSUE #204: promtool check metrics issues
Hi @oliver006 ,

Checking the redis_exporter `v0.23.0 output with promtool, I have the following result,

```
# curl -s http://fqdn:9121/metrics | promtool check metrics
redis_commands_duration_seconds_total non-counter metrics should not have "_total" suffix
redis_commands_processed_total non-counter metrics should not have "_total" suffix
redis_commands_total non-counter metrics should not have "_total" suffix
redis_connections_received_total non-counter metrics should not have "_total" suffix
redis_evicted_keys_total non-counter metrics should not have "_total" suffix
redis_expired_keys_total non-counter metrics should not have "_total" suffix
redis_keyspace_hits_total non-counter metrics should not have "_total" suffix
redis_keyspace_misses_total non-counter metrics should not have "_total" suffix
redis_net_input_bytes_total non-counter metrics should not have "_total" suffix
redis_net_output_bytes_total non-counter metrics should not have "_total" suffix
redis_rejected_connections_total non-counter metrics should not have "_total" suffix
```

If you agree, I could send a PR for fixing this, and also adding this check to CircleCI.

### PR DESCRIPTION
**Pull request by oliver006**
### Why
This project was started over four years ago and a lot has changed since. This is an effort to get the redis_exporter to more closely adhere to the [Best Practices for Writing Exporters](https://prometheus.io/docs/instrumenting/writing_exporters/) and clean up the code.

The biggest change is that the exporter no longer directly supports configuring (and scraping) multiple redis instances but pushes that responsibility to the Prometheus server. If this is how you currently use the exporter, then have a look [at this section](https://github.com/oliver006/redis_exporter/blob/oh_1.0_here_we_come/README.md#prometheus-configuration-to-scrape-several-hosts) how to configure the Prometheus server to scrape multiple instances.

However, if you already use the exporter as intended by the Prometheus project and scrape only one instance (and e.g. use the `--redis.addr` and `--redis.password` cmd line flags) then little to nothing will change for you. 

### Breaking changes in this PR:
- The exporter will no longer directly do any Service Discovery and/or scrape multiple redis instances at once.\
All this functionality should happen in the Prometheus server.<br> You can still scrape multiple redis instances with one exporter, [see the README for an example ](https://github.com/oliver006/redis_exporter/blob/oh_1.0_here_we_come/README.md#prometheus-configuration-to-scrape-several-hosts) Prometheus server configuration.
  - this means all the `addr` and `alias` labels are gone, use Prometheus server label rewriting for this if need/want them back.
  - introduced `/scrape` endpoint to allow scraping a particular instance
    - [x] fixes #126 Parallelize requests
    - [see the README for an example ](https://github.com/oliver006/redis_exporter/blob/oh_1.0_here_we_come/README.md#prometheus-configuration-to-scrape-several-hosts)
- The exporter now uses ConstMetrics to correctly export counters and gauges. as per the best practices
  - fixes #178 Export counters as counters
  - fixes #204 fix promtool linting errors
- `total_system_memory_bytes` will not be included by default any longer, use a command line flag to include it anyway.
- renamed `latency_spike_milliseconds` -> latency_spike_seconds
- renamed all the cpu metrics to end with `_seconds`
- `last_scrape_duration` and `scrape_error` metrics are now per-instance (if you scrape multiple instances) and `scrape_error` includes the error as a label value
- Tile38 metrics are only scraped if `--is-tile38` command line flag is set
- [x] #173 Put the release file in a folder inside the .tar.gz

<br>

### New features:
- [x] related to #251 - added a flag to allow skipping cert validation
- [x] #146 allow overriding CONFIG command name

<br>

### Todo
- [x] remove 1.0.0-rc drone step
- [x] fix up link to README in this PR to link to master for future reference

### CODE REVIEW

**Comment by oliver006** [comment]
man-crossing-street

**Comment by oliver006** [comment]
This is available on docker as `v1.0.0-rc.1`

`docker run -d --name redis_exporter -p 9121:9121 oliver006/redis_exporter:v1.0.0-rc.1`

**Comment by oliver006** [comment]
@SuperQ - this PR addresses the issue you raised in #228 - do you mind having a look, at the feature in particular and maybe the PR in general? Could always some more feedback here, esp if the general direction makes sense.

@brian-brazil - this PR addresses #31 that you raised a million years ago. Mind having a look if the general idea of the changes is somewhat sane?  (cc @jkohen re: issue #178 )

File: README.md [hunk 1]
```diff
@@ -19,23 +19,83 @@ Supports Redis 2.x, 3.x, 4.x, and 5.x
...
+```yaml
 
-...
+scrape_configs:
+  ## config for the multiple redis targets that the exporter will scrape
+  - job_name: 'redis_exporter_targets'
+    static_configs:
+      - targets:
+        - redis://first-redis-host:6379
+        - redis://h:password@second-redis-host:6379
```
Comment (brian-brazil) [inline README.md]: This is not a safe way to pass a password.

File: exporter/redis.go [hunk 2]
```diff
@@ -108,21 +102,12 @@ var (
...
-		"rejected_connections":       "rejected_connections_total",
-		"expired_keys":               "expired_keys_total",
-		"evicted_keys":               "evicted_keys_total",
-		"keyspace_hits":              "keyspace_hits_total",
-		"keyspace_misses":            "keyspace_misses_total",
-		"pubsub_channels":            "pubsub_channels",
-		"pubsub_patterns":            "pubsub_patterns",
-		"latest_fork_usec":           "latest_fork_usec",
+		"instantaneous_ops_per_sec": "instantaneous_ops_per_sec",
+		"instantaneous_input_kbps":  "instantaneous_input_kbps",
```
Comment (brian-brazil) [inline exporter/redis.go]: bytes/s is preferred

File: exporter/redis.go [hunk 3]
```diff
@@ -108,21 +102,12 @@ var (
...
+		"instantaneous_output_kbps": "instantaneous_output_kbps",
+		"pubsub_channels":           "pubsub_channels",
+		"pubsub_patterns":           "pubsub_patterns",
+		"latest_fork_usec":          "latest_fork_usec",
```
Comment (brian-brazil) [inline exporter/redis.go]: seconds

File: exporter/redis.go [hunk 4]
```diff
@@ -162,96 +141,62 @@ var (
...
+		"rejected_connections":   "rejected_connections_total",
+		"total_net_input_bytes":  "net_input_bytes_total",
+		"total_net_output_bytes": "net_output_bytes_total",
+
+		"expired_keys":    "expired_keys_total",
+		"evicted_keys":    "evicted_keys_total",
+		"keyspace_hits":   "keyspace_hits_total",
+		"keyspace_misses": "keyspace_misses_total",
+
+		"used_cpu_sys":           "used_cpu_sys_seconds_total",
```
Comment (brian-brazil) [inline exporter/redis.go]: Used is implied here

File: exporter/redis.go [hunk 5]
```diff
@@ -282,104 +227,146 @@ func parseKeyArg(keysArgString string) (keys []dbKeyPair, err error) {
...
-			Namespace: namespace,
-			Name:      "exporter_last_scrape_error",
-			Help:      "The last scrape error status.",
-		}),
+
+		targetScrapeDuration: prometheus.NewSummary(
+			prometheus.SummaryOpts{
+				Namespace: opts.Namespace,
+				Name:      "target_scrape_collection_duration_seconds",
+				Help:      "Duration of collections by the SNMP exporter",
```
Comment (brian-brazil) [inline exporter/redis.go]: SNMP?

File: exporter/redis.go [hunk 6]
```diff
@@ -282,104 +227,146 @@ func parseKeyArg(keysArgString string) (keys []dbKeyPair, err error) {
...
 
-	e.initGauges()
+	if opts.IncludeMetricTotalSysMemory {
+		metricMapGauges["total_system_memory"] = "total_system_memory_bytes"
+	}
+
+	e.metricDescriptions = map[string]*prometheus.Desc{}
+	e.metricDescriptions["up"] = newMetricDescr(opts.Namespace, "up", "Information about the Redis instance", nil)
+	e.metricDescriptions["instance_info"] = newMetricDescr(opts.Namespace, "instance_info", "Information about the Redis instance", []string{"role", "redis_version", "redis_build_id", "redis_mode", "os"})
+	e.metricDescriptions["last_scrape_duration"] = newMetricDescr(opts.Namespace, "exporter_last_scrape_duration_seconds", "The last scrape duration", nil)
```
Comment (brian-brazil) [inline exporter/redis.go]: You'd generally provide information about this scrape.

File: exporter/redis.go [hunk 7]
```diff
@@ -502,13 +489,26 @@ func extractConfigMetrics(config []string, addr string, alias string, scrapes ch
...
 		if val, err := strconv.ParseFloat(strVal, 64); err == nil {
-			scrapes <- scrapeResult{Name: fmt.Sprintf("config_%s", config[pos*2]), Addr: addr, Alias: alias, Value: val}
+			e.registerGaugeValue(ch, fmt.Sprintf("config_%s", config[pos*2]), val)
 		}
 	}
 	return
 }
 
-func (e *Exporter) extractTile38Metrics(info []string, addr string, alias string, scrapes chan<- scrapeResult) error {
+func (e *Exporter) registerGaugeValue(ch chan<- prometheus.Metric, metric string, val float64, labels ...string) {
```
Comment (brian-brazil) [inline exporter/redis.go]: The name of this function doesn't make it obvious that this is actually a ConstMetric.

File: exporter/redis.go [hunk 8]
```diff
@@ -552,64 +550,52 @@ func (e *Exporter) handleMetricsCommandStats(addr string, alias string, fieldKey
...
-			e.metrics["connected_slave_lag_seconds"].WithLabelValues(
-				addr,
-				alias,
-				slaveIp,
-				slavePort,
-				slaveState,
-			).Set(lag)
+	if slaveOffset, slaveIP, slavePort, slaveState, slaveLag, ok := parseConnectedSlaveString(fieldKey, fieldValue); ok {
+		e.registerGaugeValue(ch,
+			"connected_slave_offset",
```
Comment (brian-brazil) [inline exporter/redis.go]: What's the unit for this?

File: exporter/redis.go [hunk 9]
```diff
@@ -936,152 +1017,35 @@ func (e *Exporter) scrapeRedisHost(scrapes chan<- scrapeResult, addr string, idx
...
 			latencyResult := tempVal[0].([]interface{})
-			if _, err := redis.Scan(latencyResult, &eventName, &spikeLast, &milliseconds, &max); err == nil {
-				e.metricsMtx.RLock()
-				e.metrics["latency_spike_last"].WithLabelValues(addr, e.redis.Aliases[idx], eventName).Set(float64(spikeLast))
-				e.metrics["latency_spike_milliseconds"].WithLabelValues(addr, e.redis.Aliases[idx], eventName).Set(float64(milliseconds))
-				e.metricsMtx.RUnlock()
+			var spikeLast, spikeDuration, max int64
+			if _, err := redis.Scan(latencyResult, &eventName, &spikeLast, &spikeDuration, &max); err == nil {
+				spikeDuration = spikeDuration / 1e6
+				e.registerGaugeValue(ch, "latency_spike_last", float64(spikeLast), eventName)
```
Comment (brian-brazil) [inline exporter/redis.go]: timestamp_seconds usually

**Comment by oliver006** [comment]
Thanks @brian-brazil - that's god feedback.

File: exporter/redis.go (hunk 8 above)
Comment (oliver006) [reply exporter/redis.go]: I think bytes. Rename it to `connected_slave_offset_bytes`?

File: exporter/redis.go (hunk 9 above)
Comment (oliver006) [reply exporter/redis.go]: 👍 - renamed

File: exporter/redis.go (hunk 7 above)
Comment (oliver006) [reply exporter/redis.go]: Changed.

File: exporter/redis.go (hunk 6 above)
Comment (oliver006) [reply exporter/redis.go]: Ok, moved that up to the exporter level and made it a summary.

File: exporter/redis.go (hunk 5 above)
Comment (oliver006) [reply exporter/redis.go]: copy-pasta, fixed

File: exporter/redis.go (hunk 4 above)
Comment (oliver006) [reply exporter/redis.go]: Removed `used_`

File: exporter/redis.go (hunk 2 above)
Comment (oliver006) [reply exporter/redis.go]: Removed them, they can be calculated from other metrics.

File: main.go [hunk 10]
```diff
@@ -38,23 +37,19 @@ func getEnvBool(key string) (envValBool bool) {
...
+		listenAddress            = flag.String("web.listen-address", getEnv("REDIS_EXPORTER_WEB_LISTEN_ADDRESS", ":9121"), "Address to listen on for web interface and telemetry.")
+		metricPath               = flag.String("web.telemetry-path", getEnv("REDIS_EXPORTER_WEB_TELEMETRY_PATH", "/metrics"), "Path under which to expose metrics.")
+		logFormat                = flag.String("log-format", getEnv("REDIS_EXPORTER_LOG_FORMAT", "txt"), "Log format, valid options are txt and json")
+		configCommand            = flag.String("config-command", getEnv("REDIS_EXPORTER_CONFIG_COMMAND", "CONFIG"), "What to use for the CONFIG command")
+		isDebug                  = flag.Bool("debug", getEnvBool("REDIS_EXPORTER_DEBUG"), "Output verbose debug information")
+		showVersion              = flag.Bool("version", false, "Show version information and exit")
+		redisMetricsOnly         = flag.Bool("redis-only-metrics", getEnvBool("REDIS_EXPORTER_REDIS_ONLY_METRICS"), "Whether to also export go runtime metrics")
+		inclMetricTotalSysMemory = flag.Bool("incl-metric-total-system-memory", getEnvBool("REDIS_EXPORTER_INCL_METRIC_TOTAL_SYS_MEM"), "Whether to include the redis_total_system_memory_bytes metric")
```
Comment (SuperQ) [inline main.go]: How about a simple `include-system-metrics` flag.

Comment (oliver006) [reply main.go]: Yeah, I like it, let's do that.

File: exporter/redis.go (hunk 8 above)
Comment (SuperQ) [reply exporter/redis.go]: Yes, that sounds good.

Comment (oliver006) [reply exporter/redis.go]: 👍

File: README.md (hunk 1 above)
Comment (SuperQ) [reply README.md]: Agreed.

I would suggest that for `/scrape` targets that `redis://` is implied. It doesn't make sense to remotely scrape a unix socket. Only allowing `host:port`, would be a good thing to filter for in the input.

Comment (oliver006) [reply README.md]: I don't think we can simply drop the scheme from the target as some instances are using SSL and the `rediss://` scheme is used for that.

But I agree, this is not the best way to pass a password and I'm open for suggestions on how to enable users to scrape password-protected instances.

Comment (brian-brazil) [reply README.md]: What we do elsewhere is have the password come from a config file.

Comment (oliver006) [reply README.md]: > What we do elsewhere is have the password come from a config file.

That's doable what what would that look like then someone wants to scrape several password protected instances use the `/scrape` endpoint?

Comment (brian-brazil) [reply README.md]: We have different modules in the snmp/blackbox exporter for that, which is basically another url parameter.

Comment (oliver006) [reply README.md]: I'm not sure I understand what you mean.

Comment (brian-brazil) [reply README.md]: https://github.com/prometheus/snmp_exporter/blob/master/generator/FORMAT.md

Comment (oliver006) [reply README.md]: Hmm, I see. So it looks like you use the module name to indicate to the exporter which auth profile to use?
I guess I could do something similar. Have a password file deployed with the exporter where each password is associated with an ID and then the prometheus server can send the password_id as a parameter.

Comment (SuperQ) [reply README.md]: I would suggest having one password from the normal command line/env methods to start with and iterate on additional passwords. This way we can get this merged and have a good start on better functionality.

Comment (oliver006) [reply README.md]: Sounds good, let's start with that so this PR can move forward.

File: main.go [hunk 11]
```diff
@@ -145,9 +115,10 @@ func main() {
 		http.Handle(*metricPath, promhttp.Handler())
 	}
 
+	http.HandleFunc("/scrape", exp.ScrapeHandler)
```
Comment (zoonage) [inline main.go]: Isn't `/metrics` the standard endpoint? Also I think most exporters provide this path as an argument

Comment (oliver006) [reply main.go]: Good question! Yes, `/metrics` normally is the standard endpoint and that's what's still used for scraping the exporter metrics. It's configurable via `--web.telemetry-path` and hooked up in main.go line 115.
`/scrape` is what's used for scraping multiple redis instances,  [see the README here](https://github.com/oliver006/redis_exporter/blob/oh_1.0_here_we_come/README.md#prometheus-configuration-to-scrape-several-hosts).

**Comment by oliver006** [comment]
Let's do this.

**Comment by oliver006** [comment]
Github has issues with notifications right now, the exporter binaries and docker image for 1.0.0 will be available as soon as these have been resolved.

**Comment by roman-vynar** [comment]
> The exporter will no longer directly do any Service Discovery and/or scrape multiple redis instances at once.

That's very sad. We are running multiple redis containers on a single instance which means we now need to run multiple redis_exporter processes :(

We use Prometheus and we do Service Discovery but we don't want to run multiple redis_exporters :(

**Comment by SuperQ** [comment]
You don't need multiple exporters. The new version can still scrape multiple redis instances. It is configured more like the blackbox exporter.

**Comment by oliver006** [comment]
@roman-vynar - [see here](https://github.com/oliver006/redis_exporter#prometheus-configuration-to-scrape-several-hosts) for more information how to configure Prometheus to scrape multiple hosts.

**Comment by amangupta0297** [comment]
Hi,

Can anyone help me with redis exporter metrics
redis-exporter version: 1.10.0
exporter running at : 9121
grafana dashboard: 763

prometheus config:
  - job_name: 'redis_exporter'
    static_configs:
    - targets: ['DNS:9121']

target is healthy under prometheus, but still data is not coming over grafana.

When i see, the metrics

curl http://DNS:9121/metrics | grep redis


redis_exporter_build_info{build_date="2018-09-20-18:15:12",commit_sha="8bb0b841e9a70b0348f69483e58fea01d521c47a",golang_version="go1.10.4",version="v0.21.2"} 1
redis_exporter_last_scrape_duration_seconds 0.000292674
redis_exporter_last_scrape_error 1
redis_exporter_scrapes_total 125
redis_up{addr="redis://localhost:6379",alias=""} 0


issue: only very few metrics are coming, because of which data is not coming over grafana.

metrics  required:
1. redis_uptime_in_seconds
2. redis_connected_clients
3. redis_memory_used_bytes/ redis_memory_max_bytes

none of the data is coming, can anyone help. What am i missing

Thanks in advance

**Pull request merged**